src/index.ts

```diff
@@ -23,13 +39,46 @@ export {
 	Resolver,
 };
 
+export interface ResolveListeners {
+	close?: Array<() => void>;
+	data?: Array<(chunk: Buffer | string) => void>;
+	end?: Array<() => void>;
+	error?: Array<(e: Error) => void>;
+	readable?: Array<() => void>;
+	'resolved-name'?: Array<(dockerfilePath: string) => void>;
+	resolver?: Array<(resolverName: string) => void>;
+}
+
+/**
+ * Reads bundle.tarStream (typically a reference to BuildTask.buildStream) and
+ * resolves it to a Docker-compatible tar stream (i.e. selects a specified
+ * Dockerfile, translates Dockerfile.template vars and the like).
+ * @param bundle Bundle including the input tar stream
+ * @param resolvers Bundle resolvers
+ * @param resolveListeners Event listeners for tar stream resolution.
+ * Note that although you can add listeners to the returned Pack stream,
+ * the stream is piped to before it is returned and events may be missed
+ * unless the listeners are installed through the resolveListeners argument.
+ * You should always add at least an 'error' handler, or uncaught errors
+ * may crash the app.
+ * @param dockerfile User-selected Dockerfile (docker-compose `dockerfile:`)
+ */
 export function resolveInput(
 	bundle: Bundle,
 	resolvers: Resolver[],
+	resolveListeners: ResolveListeners,
 	dockerfile?: string,
-): Readable {
```

Comment:
tar.Pack implements the Readable interface, so this should be a "safe" change. Some callers of resolveInput were doing explicit cast from Pack to Readable, and this change avoids the need for that.

Reply:
It used to be that the typings were incompatible iirc, but if it's all working well now then I agree this is worth it.